[DOC] Add syntax primer to docs (#3841)
grafana/tempo: docs/sources/tempo/traceql/trace-structure.md

---
title: Trace structure and TraceQL
menuTitle: Trace structure and TraceQL
description: Learn about trace structure and TraceQL queries.
weight: 200
keywords:
  - syntax
  - TraceQL
---

# Trace structure and TraceQL

> The TraceQL language is designed to use the same syntax and semantics as PromQL and LogQL, where possible.

Reviewer: This doesn't strike me as exactly true since much of the syntax has diverged, though I'm not sure what to say in this respec.

Author: how's this? 
```suggestion
Inspired by PromQL and LogQL, TraceQL uses similar the same syntax and semantics .
```

Reviewer: Yeah, something like that reads better to me.

## Trace structure

Traces are telemetry data structured as trees.
Traces are made of spans (for example, a span tree); there is a root span that can have zero to multiple branches that are called child spans.
Each child span can itself be a parent span of one or multiple child spans, and so on so forth.

![Trace_and_spans_in_tree_structure](/media/docs/tempo/traceql/trace-tree-structures-and-spans.png)

In the specific context of TraceQL, a span has the following associated fields:

- **name**: the span name
- **duration**: difference between the end time and start time of the span
- **status**: enum: `{ok, error, unset}`. For details, refer to [OTel span status](https://opentelemetry.io/docs/concepts/signals/traces/#span-status) documentation.
- **kind**: enum: `{server, client, producer, consumer, internal, unspecified}`. For more details, refer to [OTel span kind ](https://opentelemetry.io/docs/concepts/signals/traces/#span-kind) documentation.
- Attributes

The first four properties are *intrinsics*.
They are fundamental to the span.

*Attributes* are custom span metadata in the form of key-value pairs.

> There are two types of attributes: **span attributes** and **resource attributes**.

Reviewer: This is referring to the scope of the query.  I think we have an additional scope recently as well.

Author: @zalegrala Is this additional attribute available in the 2.5 release or for the 2.6 release? Our 2.5 docs say two attributes. 

If it's the latter, we could change it here and then remove that update from the 2.5 backport.

Author: Our TraceQL doc for 2.6 says "TraceQL has four different attribute scopes: span attributes, resource attributes, event attributes, and link attributes."

Reviewer: Yep, you're right, good point.

Author: Actually, I could create a doc issue to update the attribute list for this page and also to check through the rest of the TraceQL doc to make sure that it's all up to date. That might be the better solution so everything is consistent.

Author: i'm glad you mentioned the other attributes. That's important to make sure that things are updated throughout.

Author: I've added it to the Impove TraceQL doc epic. #353

*Span attributes* are key-value pairs that contain metadata that you can use to annotate a span to carry information about the operation it's tracking.
For example, in an eCommerce application, if a span tracks an operation that adds an item to a user’s shopping cart, the user ID, added item ID and cart ID can be captured and attached to the span as span attributes.

A *Resource attribute* represents information about an entity producing the span.

> For example, a span created by a process running in a container deployed by Kubernetes could link a resource that specifies the cluster name, namespace, pod, and container name.

Reviewer: The SDK and version might also be good candidates for examples.

Author: Do you think this is something we should add in this PR or open another doc issue to add those examples? I'd like to get this content in, if we can.

Reviewer: Let's take what you've got and we can sync if we need to.

Author: I've added it as a line item to my TraceQL doc epic.

Resource attributes are resource-related metadata (key-value pairs) that are describing the Resource.

The OpenTelemetry specification defines Semantic Attributes for Spans and for Resources.
Semantic Span Attributes are a set of naming schemes for attributes shared across languages, frameworks, and runtimes.
For more details, refer to [Trace Semantic Conventions](https://opentelemetry.io/docs/specs/semconv/general/trace/) and [Resource Semantic Conventions](https://opentelemetry.io/docs/specs/semconv/resource/).

## TraceQL queries

The purpose of TraceQL is to search or query for spans.
The query returns a set of spans, also called a spanset.

A TraceQL query can select traces based on:

- span attributes, timing and duration
- structural relationships between spans
- aggregated data from the spans in a trace

As with PromQL and LogQL, the query is structured as a pipeline of operations (filters and aggregators).

> The query expression is evaluated on one trace at a time, selecting or discarding spans from the trace.

Reviewer: "from the trace", might be "from the result".

Author: Done.

> The query expression is evaluated on one trace at a time, selecting or discarding spans from the trace.

Author: ```suggestion
The query expression is evaluated on one trace at a time, selecting or discarding spans from the result.
```
